src/apps/answers/crud/answers.py

```diff
@@ -234,6 +234,17 @@ async def get_respondents_submit_dates(
         query = query.where(func.date(AnswerSchema.created_at) >= filters.from_date)
         query = query.where(func.date(AnswerSchema.created_at) <= filters.to_date)
         query = query.where(AnswerSchema.applet_id == applet_id)
+        query = query.where(
+            case(
+                (
+                    AnswerSchema.flow_history_id.isnot(None),
+                    AnswerSchema.id_from_history_id(AnswerSchema.flow_history_id) == str(filters.activity_or_flow_id),  # noqa: E501
+                ),
+                else_=AnswerSchema.id_from_history_id(AnswerSchema.activity_history_id)  # noqa: E501
+                == str(filters.activity_or_flow_id),
```

Comment:
I'm not clear why we need to convert the UUID to string. The `id_from_history_id` functions return a `UUID | none` type, so it should be fine to do quality comparison against another `UUID | none` type, no? Maybe I'm missing something…

```suggestion
                    AnswerSchema.id_from_history_id(AnswerSchema.flow_history_id) == filters.activity_or_flow_id,
                ),
                else_=AnswerSchema.id_from_history_id(AnswerSchema.activity_history_id) == filters.activity_or_flow_id,
```

Reply:
yes, but `filters.activity_or_flow_id` should be UUID, because we should validate it by pydantic, we can not accept a free format string, so original code should work, calling `str(filters.activity_or_flow_id)` will have same behavior

Reply:
please, do not use `# noqa: E501` you can break the line, use `make cqf` to automatically fix this, ignoring a lint or mypy error should be always the last solution

Reply:
> causing an error parsing as uuid, that's why I had to use string instead

@ramirlm I am not seeing any errors related to the types in my VS Code when using UUID. I'm also not seeing any errors that suggest that `# noqa: E501` is needed anywhere. With the code I suggested above, everything works fine, no issues. The reason `make cqf` isn't making any formatting changes is because there's no issue with the formatting, at least on my end.

I'm curious where you're seeing errors @ramirlm – is it in VS Code? Perhaps there's a configuration issue? 🤔

Reply:
Oh, you're talking about a runtime error. I was only talking about linting errors till now, but I've reproduced your runtime error.

The error you showed is actually the result of the `str()` coercion above, which is what I was trying to point out is not correct. The `str()` is coercing `None` (when the parameter is omitted) to a zero-length string `''`, which is not a well-formed UUID string.

When I updated the code to what I suggested above – without the `str()` coercion:
```py
                (
                    AnswerSchema.flow_history_id.isnot(None),
                    AnswerSchema.id_from_history_id(AnswerSchema.flow_history_id) == filters.activity_or_flow_id,
                ),
                else_=AnswerSchema.id_from_history_id(AnswerSchema.activity_history_id) == filters.activity_or_flow_id,
```
the error went away.

Are you seeing something different?

Reply:
ok, now I understood the problem here,
 - without str() in filters.activity_or_flow_id, we have a comparison between str resulted from AnswerSchema.id_from_history_id()  and UUID, this will not work, we should keep str()
  - We should include this where clause only when filters.activity_or_flow_id is not None, it make more sense to me, because it is a optional filter

Reply:
@rcmerlo I don't quite follow your first bullet. According to my VS Code, `AnswerSchema.id_from_history_id` returns a `UUID` or `None`, not a string:

![CleanShot 2024-12-16 at 11 27 23@2x](https://github.com/user-attachments/assets/431d4641-daf1-466f-8834-0a445a313095)

So I don't see the reason for keeping `str()`.

However, I agree we should omit this clause altogether if `activity_or_flow_id` is `None`.

Reply:
Ok, you are right, I was confused because of hybrid_method decorator from SQLAlchemy used in this method